### ISSUE TITLE
GitVersion of file path where last changed (GitVersion -filter)

### ISSUE THREAD

**Issue by o-o00o-o** [GitTools/GitVersion#570]
In the company that I currently work they prefer to have a Git repository that contains multiple subsystems and components (e.g. Website, Database etc). It simplifies a number of aspects to the development, admin, approval, build and deploy process. It also avoids heavy use of subcomponents/subtree's etc.

One issue however after we have switched to SemVer is that now the SemVer is reflecting the overall system version. That is quite a good thing as it means that we can simplify the version for our users etc. However it would also be useful to be able to identify the GitVersion of particular subsystems/components within the one repo.

This could be useful for a few reasons but the big one is deployments. If we have changed the Website but not the database for a version, we would rather skip the deployment operation. But if the database version has incremented then we would have no simple way to identify whether we should skip or perform the database deployment. Note that we could avoid this problem by not building the db but In each case we have to build all components so we can perform the appropriate unit/integration tests for each subsystem.. and also how do we know not to build even if that were an option? We lead back to the same problem that we need to get the SemVer of each subsystem independently at build time.

Note that I am suggesting that we still only have one SemVer lineage, not totally independent SemVers for each component. So if all components are changed, they will always converge back to all being on the same version number. 

How do we do it. My idea is that GitVersion already crawls through the Git repo to determine the version that should apply. So if we can supply a regex at runtime that indicates filepaths to include then we can get a "filtered GitVersion". This will then only consider commits where there was a change to any file that matches the regex.

So to get Db version we simply apply the regex "^Db/.*$" and this will always return
E.g.

Our folder structure looks like this
/Web/
/Db/

We start at v1.0.0. 
We make a change to Web only and release. System = v1.1.0 , Website = v1.1.0, Db = v1.0.0
We now make a change to the Db only and release. System = v1.2.0, Website = v1.1.0, Db = v1.2.0
We now make a breaking api change to both and release. System = v2.0.0, Website = v2.0.0, Db=v2.0.0

I guess this would be implemented via a parameter to the command line version and also for us it would be needed to be available in the MsBuild task.

So at command line 

GitVersion

 would work as today and 

GitVersion -filter "^Db\.*$" 

would return the version of the Db folder only.

I hope I have explained my idea clearly enough and hope others can see the benefit and use. 

There may well be other uses for such a filter that I haven't yet thought of.

Thanks for reading and hope this is not difficult to implement.

**Comment by JakeGinnivan** [comment]
GitVersion does not look at contents or what has changed, so technically this is not possible. But on a higher level I don't think that this is a good idea at all.

If the System, website and DB are different deployment units, they should have their own repository and their own version.

Then you could have a master repository which has system, website and DB repos as submodules. That way you keep the same folder structure, commit to each submodule separately and each submodule is versioned independently.
I am going to close this issue, but we can continue the discussion about reasons why separate repositories would not be an option?

**Comment by o-o00o-o** [comment]
Separate repo’s was my first thought also. However the Git administrators in my company are very against creating more than one repo for each System despite some discussion around this.

The reasons why multiple repo’s per system is rejected is that:

-          We have a number of independent, custom-built websites for our customers. In this case, the website, database, reports and ssis are not independent of each other, so they will never be used as true components as are designed to work exclusively together.

-          Each collection of repo’s are independent security units which makes administration more complex to setup and maintain when you want them all the be the same always.

-          The web is littered with posts on how complex things can get when using Submodules and Subtree’s with GIT. The company is new to Git but the strategy due to those posts is to avoid unless absolutely necessary. Due to the above it is seen as avoidable so they are.

-          A single feature request to a system often require changes to both website and database and this way we keep all of those items together in one branch. Also we then have just a single pull request and everything can be seen together in one place.

-          I’m not sure that it is a universal truth that the boundary for a Git repo should be defined by a deployment unit as you suggest. There is nothing else in our build/deploy pipeline that requires that and changing the whole git structure and guidelines just for an automated version control is not seen as a good enough reason. A problem is that versioning is not seen as a big problem (although to the enlightened we know it is! ☺).. but it is one of those concepts that is understood best with experience rather than through argument.

So in summary while in some cases it may make sense overall to divide up repo’s as per deployable units, I’m not sure it always is and ours is an arguable case.

It would be good if GitVersion would allow the user to choose rather than force a particular repo strategy.

The only real issue we have found with having everything in one repo is with GitVersion, hence why I thought others may be in a similar situation and why I suggested the filter approach.

Our build server bamboo is happy to deal with subsets of repo’s (with a similar regex approach as I’m suggesting here). This is where I got some inspiration for this filter switch.

GitVersion already does the hard job of telling the correct version given the other indicators in the git repo. With this switch we would just be considering some more indicators in the repo.

I understand that running with a filter command may add processing  overhead but at least this would be a choice the user can make overall to their solution. i.e. slightly slower builds etc but easier administration or faster builds but harder administration.

From my limited understanding of the details on how GitVersion works I don’t understand why it isn’t *technically* possible? I understand it currently isn’t looking at what has changed but this is what this feature would add if this switch was used.

To clarify that point - my understanding is that GitVersion is traversing the commit tree from the current commit (HEAD) back through each base commit. My theory was that as part of each check from the HEAD position back it just needs to look at the file paths that were included in each commit (I believe this is easy to do). If it matches the filter then it will stop and use this commit to establish the GitVersion as today if not it will ignore that commit and move to it’s base and repeat until it gets to a commit that has a regex match. At the point a commit with a match is found, it would simply use the GitVersion algorithm as normal from that commit point (i.e. from then on it doesn’t then care about changes to files).

I hope there would be a way we can come up with to use GitVersion to solve this problem. It would be great if GitVersion was the answer no matter what Git structure we use (voluntarily or otherwise!).

If it makes a difference I would be happy to help contribute to this feature in some way. I can code c#, test, document etc.
